### PR TITLE
Support removing inner ticks in label_outer()

### PR DESCRIPTION
**Pull request by timhoffm**
Up to now `label_outer()`, only affects tick labels, but not ticks. This introduces an optional parameter `remove_inner_ticks`, which removes the inner ticks if the corresponding tick labels are removed.

### Example / Motivation:

```
x = np.linspace(0, 2 * np.pi, 100)

fig, axs = plt.subplots(2, 2, sharex=True, sharey=True,
                        gridspec_kw=dict(hspace=0.05, wspace=0.03))

axs[0, 0].plot(x, np.sin(x))
axs[0, 1].plot(x, np.cos(x))
axs[1, 0].plot(x, -np.cos(x))
axs[1, 1].plot(x, -np.sin(x))

for ax in axs.flat:
    ax.grid(color='0.9')
    ax.label_outer()
```

image

It's quite cumbersome to remove the ticks as well (same logic as a for the tick labels).

with `ax.label_outer(remove_inner_ticks=True)` this becomes trivial

image

### CODE REVIEW

File: lib/matplotlib/tests/test_subplots.py [hunk 1]
```diff
@@ -40,6 +41,20 @@ def check_visible(axs, x_visible, y_visible):
...
+    Check that the x and y tick visibility is as specified.
+
+    Note: This only checks the tick1line, i.e. bottom / left ticks.
+    """
+    for ax, visible, in zip(axs, x_visible):
+        for tick in ax.xaxis.get_major_ticks():
+            assert tick.tick1line.get_visible() == visible
+    for ax, y_visible, in zip(axs, y_visible):
+        for tick in ax.yaxis.get_major_ticks():
+            assert tick.tick1line.get_visible() == visible
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_subplots.py:55]: I am very confused by coverage claiming this line is not run, the only thing that makes sense is if in every case where we test this `ax.yaxis.get_major_ticks()` is an empty list?

Do we need to force a draw at the top of this checker?

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -4586,18 +4586,25 @@ def get_shared_y_axes(self):
...
         """
         Only show "outer" labels and tick labels.
 
         x-labels are only kept for subplots on the last row (or first row, if
         labels are on the top side); y-labels only for subplots on the first
         column (or last column, if labels are on the right side).
+
+        Parameters
+        ----------
+        remove_inner_ticks : bool, default: False
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/axes/_base.py:4599]: we could set it to `None` for a release and warn (pass the input to silence it), but that would make it a bit annoying to support multiple versions of Matplotlib (as old versions won't even take the input.

I think if we want to flip the default behavior we should have this option available for a few releases and then go through the above dance when the back-compat story is easier for users.

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -4608,17 +4615,21 @@ def _label_outer_xaxis(self, *, check_patch):
         if not ss.is_first_row():  # Remove top label/ticklabels/offsettext.
             if label_position == "top":
                 self.set_xlabel("")
-            self.xaxis.set_tick_params(which="both", labeltop=False)
+            top_kw = {'top': False} if remove_inner_ticks else {}
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/axes/_base.py:4618]: I am 👎🏻 on this.  The old behavior was to leave the tick state alone.  Failing to ask for them to be removed does not mean you are positively asking for them to be added.

File: lib/matplotlib/axes/_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/axes/_base.py:4599]: I don't mind adding this as an option, but I pretty strongly feel it should not be the default.  I rarely use gridlines, and having the tick nearby makes it easier to quantify by eye.

**Comment by jklymak** [comment]
Note that if you call `sharex=True` and `sharey=True`, you do not need to call `ax.label_outer` to get the current behaviour.  This almost says to me that we need a standalone method `ax.tick_outer()` that does what this kwarg is doing.  You could also imagine `subplots` and friends getting a `tick_outer` kwarg.

**Comment by timhoffm** [comment]
I'll think about this a bit more. - Maybe this is a reasonable extension (If I want to configure tick label visibility, I may optionally want the same for the ticks themselves. Maybe we need something fundamentally better to configure ticks, ticklabels and spines on grid layouts.

**Comment by anntzer** [comment]
From a quick look I guess a tick_outer() method would be nice indeed.
I'll move this to draft status as @timhoffm said he wants to think more; feel free to move it back to ready status at any time.

**Comment by tacaswell** [comment]
Given 

> I'll think about this a bit more. 

I'm going to push this to 3.8 unless @timhoffm disagrees.

**Comment by jklymak** [comment]
I think vanishingly few people use label_outer - it's not that useful without sharex/y and sharex/y, which basically do what label_outer does.  I think this new feature, which is genuinely useful, is going to be hard to find.

File: doc/users/next_whats_new/label_outer.rst [hunk 4]
```diff
@@ -0,0 +1,26 @@
+Remove inner ticks in ``label_outer()``
+---------------------------------------
+Up to now, ``label_outer()`` has only removed the ticklabels. The ticks lines
+were left visible. This is now configurable through a new parameter
+``label_outer(remove_inner_ticks=True)``.
+
+
+.. plot::
```
Comment (jklymak) [inline doc/users/next_whats_new/label_outer.rst:8]: This probably needs an example in the docs somewhere as well?    Perhaps also mention at https://matplotlib.org/devdocs/users/explain/axes/axes_ticks.html#styling-ticks-tick-parameters?

Comment (timhoffm) [reply doc/users/next_whats_new/label_outer.rst:8]: I plan to use it in #24379

**Review by jklymak — APPROVED** [review]
I don't think this is very discoverable, but doesn't hurt anything.

**Comment by timhoffm** [comment]
Something is very broken here on GitHub. In contrast to what the page is claiming I did not close this PR, and the commit is also not on master.

grafik

**Pull request merged**